### ISSUE TITLE
Simplify dataset definitions, loading, and source adapters

### ISSUE THREAD

**Issue by calebrob6** [torchgeo/torchgeo-bench#381]
## Goal

Review and simplify the dataset layer for readability, sensible abstractions, and long-term maintainability. Backward compatibility with the current APIs and implementation is not a constraint for this redesign. Preserve the scientific meaning of the benchmarks: dataset identities, splits, labels, sensor values, and the relationship between image channels and their metadata.

This follows the simplification stack ending at #376. The audit used [f128c65](https://github.com/torchgeo/torchgeo-bench/tree/f128c65feec33a2c4fcb243995afd56a0c2c13e6); reconcile the findings with current main before implementation.

The current layer mixes dataset definitions, storage selection, upstream compatibility, sample conversion, band selection, resizing, and batching. The largest simplifications require clearer ownership between those responsibilities.

## Concrete correctness finding: channel order can disagree with metadata

The model's band metadata is resolved independently of the actual image produced by the backend adapter. A channel-count check does not detect an ordering mismatch.

For KuroSiwo, requesting `("dem", "vv")` produces metadata in DEM/VV order, but its canonicalizer always joins post-event SAR before DEM:

```python
import torch
from torchgeo_bench.datasets.kuro_siwo import KuroSiwo

bench = KuroSiwo()
requested = ("dem", "vv")
metadata_order = [band.name for band in bench.resolve_band_specs(requested)]
sample = bench.canonicalize_sample({
    "image_dem": torch.full((1, 1, 1), 10.0),
    "image_post": torch.full((1, 1, 1), 20.0),
})

print(metadata_order)  # ["dem", "vv"]
print(sample["image"][:, 0, 0].tolist())  # [20.0, 10.0]: VV, then DEM
```

This was reproduced with synthetic tensors, without loading real dataset files. The relevant code is [`KuroSiwo.canonicalize_sample`](https://github.com/torchgeo/torchgeo-bench/blob/f128c65feec33a2c4fcb243995afd56a0c2c13e6/src/torchgeo_bench/datasets/kuro_siwo.py#L42) and [`_V2Dataset.build_band_order`](https://github.com/torchgeo/torchgeo-bench/blob/f128c65feec33a2c4fcb243995afd56a0c2c13e6/src/torchgeo_bench/datasets/geobench_v2.py#L152).

There is a related loss of interleaving in sensor grouping: TreeSatAI selection `("red", "b04", "green")` has requested source order `["red", "B04", "green"]`, while flattening the grouped upstream request gives `["red", "green", "B04"]`.

The loaded image and its `BandSpec` sequence must come from one resolved input description. Otherwise normalization and model band mapping can use the wrong metadata even when tensor dimensions look correct. Address this independently of the larger architectural cleanup.

## Recommended simplifications

Paths below omit the `src/torchgeo_bench/` prefix.

### 1. One dataset definition and one catalog

Identity and routing are distributed across `datasets/loading.py::_REGISTRY_SPEC`, `datasets/__init__.py::_LAZY_CLASSES` and `__all__`, `datasets/geobench_v2.py::_V2_REGISTRY`, wrapper class attributes, and download catalogs in `download.py`.

Introduce a lightweight immutable `DatasetSpec` describing metadata, task/target semantics, split policy, supported inputs, and source information. Catalog discovery, download routing, and loading should consume that definition. Most metadata-only subclasses could become records.

Keep per-dataset definition files so statistics and dataset-specific assumptions remain easy to find. Metadata discovery should not need to import Torch or upstream loader implementations.

### 2. Separate split loading from batching

`datasets/loading.py::get_datasets` returns either three or four positional values. It unconditionally constructs train, validation, and test datasets and dataloaders, including when `return_val=False`. Standalone profiling needs one training batch but still invokes this whole path.

Provide an explicit `load_split()` operation with one typed result. Let callers request only the splits they need, and construct dataloaders separately. Remove the redundant positional return of the training dataset alongside its loader.

A loaded split should carry the actual dataset and its resolved input metadata, avoiding repeated band-resolution decisions in model construction.

### 3. Establish a canonical sample boundary

`datasets/geobench_v2.py::_ChainedTransform` handles either `image` or arbitrary `image_*` keys. Individual wrappers also choose acquisitions, concatenate modalities, and remap masks.

Each source adapter should produce a defined sample: raw `float32` imagery with explicit channel/time axes and a target matching the dataset's classification, multilabel, or segmentation semantics. The ordered band metadata must describe that image. Common resizing and framework transforms run after this conversion and should not guess upstream key conventions.

Preserve genuine dataset-specific rules, including FOTW's selected acquisition, KuroSiwo's SAR/DEM handling, SpaceNet label remapping, and categorical-mask interpolation.

### 4. Make capabilities explicit and simplify backend construction

The V2 path currently involves a metadata subclass, `_V2Dataset`, a mostly forwarding `GeoBenchv2` dataset, and the upstream dataset. PASTIS overrides the family loading method to add temporal options instead of reusing one construction path.

Use small backend-specific factories/adapters and explicit temporal, split, and source-band options. PASTIS should supply temporal settings without duplicating common construction. Avoid leaking untyped `**kwargs` and `band_order: object` through the public dataset interface.

Unsupported options should fail before I/O with clear errors. Currently unsupported partitions are ignored with a warning, while `time_steps` is dynamically forwarded to methods that do not accept it. Keep default downloading behavior explicit in the common backend path; the inspected upstream PASTIS default is currently `download=False`, so its omission is a maintenance risk rather than a confirmed automatic-download bug.

### 5. Prefer one canonical V1 runtime format

`datasets/geobench_v1.py::_V1Dataset.get_dataset` chooses shards or custom JSON-metadata HDF5 based on which directories exist. The two readers repeat partition handling, band-name matching, label conversion, and sample construction. `data_root()` points to the HDF5 layout even though shards are preferred.

With compatibility out of scope, prefer verified shards as the sole V1 benchmark runtime format. Keep HDF5 conversion outside the runtime if that import workflow is still useful. This removes filesystem-dependent backend selection and duplicated decoding logic.

Separate downloading/verification from local reading. Preserve JSON-only metadata, pinned V1 downloads, archive checksums, and visible failures for corrupt data.

### 6. Give band sets accurate names

`rgb_bands` sometimes means a dataset-specific reduced/default input: CaFFe supplies one grayscale channel, and KuroSiwo supplies two SAR channels.

Separate the default band set from an actual RGB band set. An explicit default selection can work across datasets without calling grayscale or SAR imagery RGB. Keep the chosen bands, ordering, and model compatibility visible.

## Illustrative target API

This is a design sketch, not an existing API:

```python
spec = get_dataset_spec("pastis")
train = load_split(spec, "train", bands="default", time_steps=4)
loader = DataLoader(train.dataset, batch_size=64, shuffle=True)
```

`train.bands` describes exactly the channels emitted by `train.dataset`. Source-specific readers stay behind the loading boundary; batching is a caller decision.

## What should remain

- Small, discoverable per-dataset metadata definitions and immutable band metadata.
- Distinct benchmark identities and statistics even when datasets share underlying imagery, such as GeoBench EuroSAT and torchgeo EuroSAT.
- Necessary source-specific readers and scientific sample conversions.
- Model-owned normalization: dataset readers emit raw sensor values.
- Explicit downloads, integrity verification, meaningful missing/corrupt-data errors, and nearest-neighbor resizing of categorical masks.

The objective is a smaller set of clear concepts, not one large registry containing every statistic or a generic loader full of backend-specific conditionals.

## Delivery and acceptance

- Fix and cover the channel-order mismatch first. Use distinguishable synthetic channel values and verify values and metadata together, including interleaved sensors and DEM-before-SAR selections.
- Redesign split loading and metadata ownership, then consolidate catalog definitions and simplify backend construction in independently reviewable changes.
- Verify that loading one split does not require or construct unrelated splits.
- Exercise single-label, multilabel, segmentation, temporal, grayscale, and multi-sensor samples against an explicit output contract.
- Validate supported options before accessing data, and retain lightweight metadata/catalog queries.
- Update callers, download dispatch, documentation, and tests together; compatibility wrappers are not required merely to preserve the old API.
- Remove retired registries, forwarding layers, duplicate conversion paths, and obsolete format-selection branches once callers have migrated.
